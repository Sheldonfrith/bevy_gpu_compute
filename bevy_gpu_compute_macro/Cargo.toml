[package]
name = "bevy_gpu_compute_macro"
version.workspace = true
authors.workspace = true
rust-version.workspace =  true
description.workspace = true
edition.workspace = true
repository.workspace = true
license.workspace = true
keywords.workspace = true
categories.workspace = true

[lib]
proc-macro = true
[dependencies]
proc-macro-error = "1.0.4"
proc-macro2 = "1.0.92"
quote = "1.0.38"
syn = { version = "2.0.94", features = ["full","visit-mut","visit"]   }
bevy_gpu_compute_core = {path = "../bevy_gpu_compute_core", version ="0.1"}
regex = "1.11.1"
bytemuck = {version = "1.21.0", features=["derive"]}


[dev-dependencies]
trybuild = "1.0.101"
pretty_assertions = "1.4.1"
