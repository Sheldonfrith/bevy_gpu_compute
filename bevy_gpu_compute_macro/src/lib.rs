#![feature(allocator_api)]

use pipeline::lib::CompilerPipeline;
use proc_macro::TokenStream;
use proc_macro_error::{proc_macro_error, set_dummy};
use syn::parse_macro_input;
mod pipeline;
/**
## *Please read this documentation carefully, especially if you are getting errors that you don't understand!*

*...because it's currently impossible with rust proc-macros to override some of the error messages, so a lot of them don't actually indicate correctly what your issue is. So this documentation is how you are going to solve them!*

Here are some pointers:
- No let statements allowed except within functions. If you want to define a variable use "const" instead.
- If you see an error like `cannot transmute between types of different sizes...`, it is probably a memory padding/alignment issue. The f16 (aka PodF16) is a common culprit. If you have structs with multiple fields of different types, try adding various amounts of padding to the structs to see if that fixes the issue.
- When accessing special WGSL types like `Vec3`, `Mat3x4`, etc. you CANNOT use parenthesis when accessing the fields. For example:
### Valid:
```rust
# use bevy_gpu_compute_core::wgsl_helpers::*;
let my_vec3 = Vec2I32::new(1,2);
// valid:
let x = my_vec3.x;
let x = my_vec3[0];
```
### Invalid:
```compile_fail
// invalid:
let x = my_vec3.x();
```
- "rgba" field access on vectors is not supported right now, use index or xyzw instead.
- Every Input/Output you want to transfer between the CPU and GPU must have its type defined within the shader module. Here's how you do that:
    - Input Vec/Array/Matrices: Define the inner-type, and put `#[vec_input]` above the type definition. Example: If you want to work with an input equivalent to `Vec<{x:f32, y:f32}>` in your module, then write
    ```ignore
    #[vec_input]
    pub struct MyStruct { x: f32, y: f32 }
    ```

    We wrap the inner type in an array for you automatically, so that you don't have to worry about data length or alignments.
* You cannot use rust's struct-literal syntax for initializing structs from the helper types module. You must use the `new` method instead. For example:
    ## Wrong:
     ```compile_fail
         #use bevy_gpu_compute_corepu_compute_core::wgsl_in_rust_helpers::*;
          const MY_CONST: Vec3Bool = Vec3Bool {
        x: true,
        y: false,
        z: true,
    };```


    ## Correct:
    ```
    #use bevy_gpu_compute_corepu_compute_core::wgsl_in_rust_helpers::*;
    const MY_CONST: Vec3Bool = Vec3Bool::new(true, false, true);
    ```
* If you see the error `the trait bound `bool: Pod` is not satisfied...` make sure you are not trying to use a `bool` in any input data or output data. The `bool` type CAN be used but only ON the GPU, it cannot be passed between the CPU and GPU.
 */
#[proc_macro_attribute]
#[proc_macro_error]
pub fn wgsl_shader_module(_attr: TokenStream, item: TokenStream) -> TokenStream {
    set_dummy(item.clone().into());
    let module = parse_macro_input!(item as syn::ItemMod);
    let compiler_pipeline = CompilerPipeline::default();
    compiler_pipeline.compile(module, true).into()
}

#[proc_macro_attribute]
#[proc_macro_error]
pub fn wgsl_shader_module_library(_attr: TokenStream, item: TokenStream) -> TokenStream {
    set_dummy(item.clone().into());
    let module = parse_macro_input!(item as syn::ItemMod);
    let compiler_pipeline = CompilerPipeline::default();
    compiler_pipeline.compile(module, false).into()
}

/// used to help this library figure out what to do with user-defined types
#[proc_macro_attribute]
#[proc_macro_error]
pub fn wgsl_config(_attr: TokenStream, item: TokenStream) -> TokenStream {
    return item;
}
/// used to help this library figure out what to do with user-defined types
#[proc_macro_attribute]
#[proc_macro_error]
pub fn wgsl_input_array(_attr: TokenStream, item: TokenStream) -> TokenStream {
    return item;
}
/// used to help this library figure out what to do with user-defined types
#[proc_macro_attribute]
#[proc_macro_error]
pub fn wgsl_output_vec(_attr: TokenStream, item: TokenStream) -> TokenStream {
    return item;
}
/// used to help this library figure out what to do with user-defined types
#[proc_macro_attribute]
#[proc_macro_error]
pub fn wgsl_output_array(_attr: TokenStream, item: TokenStream) -> TokenStream {
    return item;
}
