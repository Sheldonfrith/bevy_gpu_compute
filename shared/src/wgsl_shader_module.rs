use crate::{
    wgsl_components::{
        WORKGROUP_SIZE_X_VAR_NAME, WORKGROUP_SIZE_Y_VAR_NAME, WORKGROUP_SIZE_Z_VAR_NAME,
        WgpuShaderType, WgslShaderModuleComponent,
    },
    wgsl_shader_module_lib_portion::WgslShaderModuleLibraryPortion,
};

use super::{
    wgsl_components::{
        WgslConstAssignment, WgslFunction, WgslShaderModuleUserPortion, WgslType,
        WgslWorkgroupDeclaration,
    },
    wgsl_wgpu_binding::WgslWgpuBinding,
};

#[derive(Debug, Clone, PartialEq, Hash, Copy)]
pub enum IterSpaceDimmension {
    OneD,
    TwoD,
    ThreeD,
}
// impl as usize
impl From<IterSpaceDimmension> for usize {
    fn from(dim: IterSpaceDimmension) -> usize {
        match dim {
            IterSpaceDimmension::OneD => 1,
            IterSpaceDimmension::TwoD => 2,
            IterSpaceDimmension::ThreeD => 3,
        }
    }
}

pub struct WgslShaderModule {
    pub user_portion: WgslShaderModuleUserPortion,
    pub library_portion: WgslShaderModuleLibraryPortion,
}
impl WgslShaderModule {
    pub fn new(module: WgslShaderModuleUserPortion) -> WgslShaderModule {
        let library_portion = WgslShaderModuleLibraryPortion::from(&module);
        WgslShaderModule {
            user_portion: module,
            library_portion: library_portion,
        }
    }
    pub fn wgsl_code(&self, iter_space_dimmensions: IterSpaceDimmension) -> String {
        let mut wgsl: String = String::new();
        // first add user static consts
        self.user_portion
            .static_consts
            .iter()
            .for_each(|c| wgsl.push_str_w_newline(&c.code.wgsl_code.clone()));
        // then add any miscelanious user helper types which are internal to the GPU only, not transfered to or from th CPU
        self.user_portion.helper_types.iter().for_each(|t| {
            wgsl.push_str_w_newline(&t.code.wgsl_code.clone());
        });
        // then add library pipeline consts
        // these include lengths of arrays, and workgroup sizes
        self.library_portion.pipeline_consts.iter().for_each(|c| {
            wgsl.push_str_w_newline(&c.code.wgsl_code.clone());
        });
        // then add user uniform definitions
        self.user_portion.uniforms.iter().for_each(|u| {
            wgsl.push_str_w_newline(&u.code.wgsl_code.clone());
        });
        // then add library uniform definitions
        self.library_portion.uniforms.iter().for_each(|u| {
            wgsl.push_str_w_newline(&u.code.wgsl_code.clone());
        });
        // then add user input array definitions
        self.user_portion.input_arrays.iter().for_each(|a| {
            wgsl.push_str_w_newline(&a.item_type.code.wgsl_code.clone());
        });
        // then add user output array definitions
        self.user_portion.output_arrays.iter().for_each(|a| {
            wgsl.push_str_w_newline(&a.item_type.code.wgsl_code.clone());
        });
        // now add wgpu bindings
        self.library_portion.bindings.iter().for_each(|b| {
            wgsl.push_str_w_newline(&b.to_string());
        });
        // now add user helper functions
        self.user_portion.helper_functions.iter().for_each(|f| {
            wgsl.push_str_w_newline(&f.code.wgsl_code.clone());
        });
        // now add library helper functions
        self.library_portion.helper_functions.iter().for_each(|f| {
            wgsl.push_str_w_newline(&f.code.wgsl_code.clone());
        });
        // now add the main function
        if iter_space_dimmensions == IterSpaceDimmension::OneD {
            wgsl.push_str_w_newline("@compute @workgroup_size(64, 1, 1)");
        } else if iter_space_dimmensions == IterSpaceDimmension::TwoD {
            wgsl.push_str_w_newline("@compute @workgroup_size(8, 8, 1)");
        } else {
            wgsl.push_str_w_newline("@compute @workgroup_size(4, 4, 4)");
        }
        wgsl.push_str_w_newline(
            &self
                .user_portion
                .main_function
                .as_ref()
                .unwrap()
                .code
                .wgsl_code
                .clone(),
        );
        wgsl
    }
}

// implement push_str_w_newline for String
trait PushStrWNewline {
    fn push_str_w_newline(&mut self, s: &str);
}
impl PushStrWNewline for String {
    fn push_str_w_newline(&mut self, s: &str) {
        self.push_str(s);
        self.push_str("\n");
    }
}
