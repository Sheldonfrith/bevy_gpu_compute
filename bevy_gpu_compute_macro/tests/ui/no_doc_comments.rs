use bevy_gpu_compute_macro::wgsl_shader_module;

#[wgsl_shader_module]
pub mod something {

    /** This is doc comment */
    /// Doc comment\n
    fn main() {
        // This is a line comment
    }
}
