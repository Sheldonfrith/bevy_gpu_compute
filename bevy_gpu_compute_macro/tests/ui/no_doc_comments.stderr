error: Doc comments are not allowed in wgsl_to_rust shader modules. Please remove them.
 --> tests/ui/no_doc_comments.rs:6:5
  |
6 |     /** This is doc comment */
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0601]: `main` function not found in crate `$CRATE`
  --> tests/ui/no_doc_comments.rs:11:2
   |
11 | }
   |  ^ the main function must be defined at the crate level (in `$DIR/tests/ui/no_doc_comments.rs`)
   |
note: here is a function named `main`
  --> tests/ui/no_doc_comments.rs:8:5
   |
8  |     fn main() {
   |     ^^^^^^^^^
   = note: you have one or more functions named `main` not defined at the crate level
   = help: consider moving the `main` function definitions
